retrieval/discovery/file.go

```diff
@@ -0,0 +1,250 @@
+		// clause to always catch done signals.
+		select {
+		case <-fd.done:
+			return
+		default:
+			select {
+			case event := <-fd.watcher.Events:
+				// fsnotify sometimes sends a bunch of events without name or operation.
+				// It's unclear what they are and why they are sent - filter them out.
+				if len(event.Name) == 0 {
+					break
+				}
+				// Everything but a chmod requires rereading.
+				if event.Op^fsnotify.Chmod == 0 {
+					break
+				}
+				// Changes to a file can spawn various sequences of events with
+				// different combinations of operations. For all practical purposes
+				// this is inaccurate.
+				// The most reliable solution is to reload everything if anything happens.
+				fd.refresh(ch)
+
+			case <-ticker.C:
+				// Setting a new watch after an update might fail. Make sure we don't lose
+				// those files forever.
+				fd.refresh(ch)
+
+			case err := <-fd.watcher.Errors:
+				if err != nil {
+					glog.Errorf("Error on file watch: %s", err)
+				}
+
+			case <-fd.done:
+				return
+			}
+		}
+	}
+}
+
+// refresh reads all files matching the discoveries patterns and sends the respective
+// updated target groups through the channel.
+func (fd *FileDiscovery) refresh(ch chan<- *config.TargetGroup) {
+	ref := map[string]int{}
+	for _, p := range fd.listFiles() {
+		tgroups, err := readFile(p)
+		if err != nil {
+			glog.Errorf("Error reading file %q: %s", p, err)
+			// Prevent deletion down below.
+			ref[p] = fd.lastRefresh[p]
+			continue
+		}
+		for _, tg := range tgroups {
+			ch <- tg
+		}
+		ref[p] = len(tgroups)
+	}
+	// Send empty updates for sources that disappeared.
+	for f, n := range fd.lastRefresh {
+		m, ok := ref[f]
+		if !ok || n > m {
+			for i := m; i < n; i++ {
+				ch <- &config.TargetGroup{Source: fileSource(f, i)}
+			}
+		}
+	}
+	fd.lastRefresh = ref
+
+	fd.watchFiles()
+}
+
+// fileSource returns a source ID for the i-th target group in the file.
+func fileSource(filename string, i int) string {
+	return fmt.Sprintf("file:%s:%d", filename, i)
+}
+
+// Stop implements the TargetProvider interface.
+func (fd *FileDiscovery) Stop() {
+	glog.V(1).Info("Stopping file discovery for %s...", fd.paths)
+
+	fd.done <- struct{}{}
+	// Closing the watcher will deadlock unless all events and errors are drained.
+	go func() {
+		for {
+			select {
+			case <-fd.watcher.Errors:
+			case <-fd.watcher.Events:
+				// Drain all events and errors.
+			case <-fd.done:
+				return
+			}
+		}
+	}()
+	fd.watcher.Close()
+
+	fd.done <- struct{}{}
+
+	glog.V(1).Info("File discovery for %s stopped.", fd.paths)
+}
+
+// readFile reads a list json or protbuf targets groups from the file, depending on its
+// file extension. It returns full configuration target groups.
+func readFile(filename string) ([]*config.TargetGroup, error) {
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	var targetGroups []*config.TargetGroup
+
+	switch ext := filepath.Ext(filename); ext {
```

Comment:
The config regex only accepts lowercase - so that needs to be changed as well.
Do you agree with only accepting .yml, .yaml, and .json (case-agnostic)?